Allows Summon Magic spellbooks to be upgraded for higher level spells

cacogen opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

### Gameplay
- Summon magic spellbooks can now give different levels of spell from 0 to 4, same as the wizard's spellbook
- Higher level spellbooks can be obtained using charge, through the improved casting event or a lucky spawn
- Used spellbooks upgraded through charge or improved casting become reusable
- Charge has a 50% chance of failure or 0% if it isn't a one use book (so varedited, same as it was to make them read faster)
- 2% chance (specialist magic probability) that summon magic spawned spellbooks upgrade to level 4 on spawn, the other levels range from 10% - 50%
- Spellbooks are now affected by the improved casting summon event, which causes them to increase 1-2 levels each time
- Using a higher levelled spellbook while knowing a lower levelled version of the spell upgrades the spell
- Using a lower levelled spellbook while knowing a higher version does nothing

### Other
- Creates a specific spell naming proc, spell level setting proc and spell levelling proc to avoid copypasted code
- Changes improved casting summon event and wizard spellbook to use above procs, removing copypasted code
- Changes spellbook to one word in files to be consistent
- Removes duplicated code in lesser smoke spell
- Renames the lesser "Smoke" spell to "Lesser Smoke" to distinguish it (it's possible to get them both at once)

<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game

- Admins can spawn spellbooks and edit their variables to give people higher levelled spells
- Crew can upgrade spellbooks to get higher levelled spells, which previously only wizards had access to
- Cleans up some messy and copypasted code

<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

## Changelog
:cl: cacogen
add: Summon Magic spellbooks can now give different levels of spell (0 is default, up to 4)
add: The Charge spell has a 50% chance of upgrading a spellbook one level
add: The Improved Casting Summon Event upgrades spellbooks 1-2 levels each time
add: 2% chance (specialist magic probability) of spellbooks becoming level 4 on spawn, lower levels range between 10% - 50% (and you aren't guaranteed to get a book in the first place)
add: Used spellbooks become reusable when upgraded
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

Commit:
Give spellbooks levels

Commit:
fixed a couple of mistakes

Shadowflame909 commented:
Wait. This whole level thing unironically feels fishy.

Is this about the "Lesser Barnyard" PR not getting merged?

cacogen commented:
It's an idea I got after that was turned down. I didn't want the code I'd written to go to waste, so I added on to it to create this.

PKPenguin321 commented:
Glad this was saved since it isn't a bad idea on it's own

Commit:
adds 2% chance of summon magic spellbooks being upgraded to level 4, adds spellbook levelling to improved casting event

FantasticFwoosh commented:
But the spell event drops a single-use charge book with unlimited amount of set uses therefore probably creating a raging mages situation when survivors are also given unlimited access to books by passing replenished ones around. (not counting replenishing all energy weapons, wands etc.) :-1: 

I would be happier if the librarians back-room desk could upgrade books on a long cooldown & the charge book was removed from the availible drops, so that the snowball is contained between actual wizards and apprentices during these rounds recovering books to expand their power.

cacogen commented:
Currently there's an 80% chance of charging failing which destroys the book.

Shadowflame909 commented:
What is the proper use of this I wonder? I find the consequences way too high for the upsides.

Would it be a wizard in his den spamming charge on a spellbook.

If its that easy to dust your books trying to make them a higher level. I think this feature is gonna get forgotten about.

HomocidlePancakes commented:
Have books explode instead of being dusted, wiz is supposed to be W A C K Y afterall

ExcessiveUseOfCobblestone commented:
Nice feature but god do I hate summon X

Commit:
decreases charge failure chance to 50%, gives a 50% chance of improved casting improving spellbooks 2 levels (because in testing it's so rare, even with maximum event intensity)

cacogen commented:
>What is the proper use of this I wonder? I find the consequences way too high for the upsides.

I wasn't thinking any further than it would be fun for the crew. How much does the wizard benefit from summon magic now?

>If its that easy to dust your books trying to make them a higher level. I think this feature is gonna get forgotten about.

I agree. I'm changing it to 50%. That's still too high in my opinion but I guess I'd like to err on the side of caution. I don't think it's that likely you'll see improved spellbooks right now given the chance of getting a level 4 spellbook from the get-go is ridiculously small and improved casting isn't that common even with summon events turned up to their highest intensity (one a minute, or five casts, which would prevent summoning magic anyway). I'll think about increasing the rate of improved spellbooks spawning by default and the success rate of charge.

Commit:
stops books levelling past maximum spell level, removes barnyard stuff, adds chances for books to spawn levelled

ATH1909 commented:
> But the spell event drops a single-use charge book with unlimited amount of set uses therefore probably creating a raging mages situation when survivors are also given unlimited access to books by passing replenished ones around. (not counting replenishing all energy weapons, wands etc.) 👎
> 
> I would be happier if the librarians back-room desk could upgrade books on a long cooldown & the charge book was removed from the availible drops, so that the snowball is contained between actual wizards and apprentices during these rounds recovering books to expand their power.

You do realize that if the 80% chance for failure is rolled, the book is destroyed, right? You can't just keep rerolling on the same book until you successfully recharge it.

wesoda25 commented:
Lesser timeswap that really slows stuff down in it for a bit but doesn't actually stop time when.

Eaglendia commented:
> Have books explode instead of being dusted, wiz is supposed to be W A C K Y afterall
A lot of the syntax and style of item fortification and magical item recharging pulls from NetHack and other similar roguelikes; the Wizard even has a Wizard of Yendor showcase in his den, and items that are overcharged with fortification scrolls turn to dust in a manner reminiscent of (if not directly referential to) NetHack.

cacogen commented on code/modules/events/wizard/aid.dm:
```diff
@@ -34,25 +34,7 @@
+				S.UpdateSpellLevel(S.spell_level + 1)
 			to_chat(L, "<span class='notice'>You suddenly feel more competent with your casting!</span>")
+	for(var/obj/item/book/granter/spell/B in GLOB.spellbooks)
+		B.level_up_book(pick(1,2), TRUE)
```
this doesn't work for some reason

cacogen commented on code/game/objects/items/granters.dm:
```diff
@@ -115,16 +115,50 @@
+	qdel(S)
+
+/obj/item/book/granter/spell/proc/level_up_book(amount, message)
+	if(spell_level + amount > spell.level_max)
```
this doesn't do anything. i don't think it can get level_max without creating a new spell to get it from

cacogen commented on code/modules/spells/spell_types/rightandwrong.dm:
```diff
@@ -146,6 +146,19 @@ GLOBAL_VAR_INIT(summon_magic_triggered, FALSE)
 	var/obj/item/M = new magic_type(get_turf(H))
 	playsound(get_turf(H),'sound/magic/summon_magic.ogg', 50, 1)
 
+	if(M.type == /obj/item/book/granter/spell)
```
this needs to use istype() but even doing that the level of spellbook is not updated when it spawns

ExcessiveUseOfCobblestone commented:
spell is not a reference, it's just a type. You can look at the default values by doing `initial(spell.var)` but you can't edit it.

SpaceManiac commented:
Judging by the review notes that you left for yourself it seems that this is a work in progress. I'm therefore marking it stale due to being not updated in almost a week. It may be closed if it continues to go unupdated. Reply here if this is in error, or if the PR is closed and you want it reopened.
